Exception with Test and $instance entity

Exception in Test -s luisV3 when testing with the attached model and unit tests.

 Test Luis model2s
##[error]Process completed with exit code 134.
Run dotnet nlu test -s luisV3 -u $UNIT_TEST_FILE -o results.json
info: NLU.DevOps.CommandLine.Test.TestCommand[0]
      Running tests against NLU model...

Unhandled Exception: System.AggregateException: One or more errors occurred. (The given key '$instance' was not present in the dictionary.) ---> System.Collections.Generic.KeyNotFoundException: The given key '$instance' was not present in the dictionary.
   at System.Collections.Generic.Dictionary`2.get_Item(TKey key)
   at NLU.DevOps.Luis.LuisNLUTestClient.GetEntities(String utterance, IDictionary`2 entities) in D:\a\1\s\src\NLU.DevOps.LuisV3\LuisNLUTestClient.cs:line 151
   at NLU.DevOps.Luis.LuisNLUTestClient.LuisResultToLabeledUtterance(SpeechPredictionResponse speechPredictionResponse) in D:\a\1\s\src\NLU.DevOps.LuisV3\LuisNLUTestClient.cs:line 205
   at NLU.DevOps.Luis.LuisNLUTestClient.TestAsync(JToken query, CancellationToken cancellationToken) in D:\a\1\s\src\NLU.DevOps.LuisV3\LuisNLUTestClient.cs:line 54
   at NLU.DevOps.CommandLine.EnumerableExtensions.<>c__DisplayClass0_0`2.<<SelectAsync>g__selectWithIndexAsync|1>d.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
   at NLU.DevOps.CommandLine.EnumerableExtensions.SelectAsync[T,TResult](IEnumerable`1 items, Func`2 selector, Int32 degreeOfParallelism) in D:\a\1\s\src\NLU.DevOps.CommandLine\EnumerableExtensions.cs:line 36
   at NLU.DevOps.CommandLine.Test.TestCommand.RunAsync() in D:\a\1\s\src\NLU.DevOps.CommandLine\Test\TestCommand.cs:line 51
   --- End of inner exception stack trace ---
   at System.Threading.Tasks.Task.Wait(Int32 millisecondsTimeout, CancellationToken cancellationToken)
   at NLU.DevOps.CommandLine.Test.TestCommand.Main() in D:\a\1\s\src\NLU.DevOps.CommandLine\Test\TestCommand.cs:line 29
   at NLU.DevOps.CommandLine.Program.Run(ICommand command) in D:\a\1\s\src\NLU.DevOps.CommandLine\Program.cs:line 36
   at NLU.DevOps.CommandLine.Program.Main(String[] args) in D:\a\1\s\src\NLU.DevOps.CommandLine\Program.cs:line 18
##[error]Process completed with exit code 134.

[unittests.zip](https://github.com/microsoft/NLU.DevOps/files/4342994/unittests.zip)

--- a/src/NLU.DevOps.LuisV3.Tests/LuisNLUTestClientTests.cs
+++ b/src/NLU.DevOps.LuisV3.Tests/LuisNLUTestClientTests.cs
@@ -437,6 +437,35 @@ namespace NLU.DevOps.Luis.Tests
             }
         }
 
+        [Test]
+        public static void ThrowsInvalidOperationWhenMissingMetadata()
+        {
+            var test = "foo";
+            var builder = new LuisNLUTestClientBuilder();
+            builder.LuisTestClientMock
+                .Setup(luis => luis.QueryAsync(
+                    It.Is<PredictionRequest>(query => query.Query == test),
+                    It.IsAny<CancellationToken>()))
+                .Returns(() => Task.FromResult(new PredictionResponse
+                {
+                    Query = test,
+                    Prediction = new Prediction
+                    {
+                        TopIntent = "intent",
+                        Entities = new Dictionary<string, object>
+                        {
+                            { "entityType", new JArray { 42 } },
+                        },
+                    },
+                }));
+
+            using (var luis = builder.Build())
+            {
+                Func<Task> testAsync = () => luis.TestAsync(test);
+                testAsync.Should().Throw<InvalidOperationException>();
+            }
+        }
+
         private static IDictionary<string, object> ToEntityDictionary(IEnumerable<EntityModel> entities)
         {
             var result = new Dictionary<string, object>();

--- a/src/NLU.DevOps.LuisV3/LuisNLUTestClient.cs
+++ b/src/NLU.DevOps.LuisV3/LuisNLUTestClient.cs
@@ -97,7 +97,7 @@ namespace NLU.DevOps.Luis
 
         private static IEnumerable<Entity> GetEntities(string utterance, IDictionary<string, object> entities)
         {
-            if (entities == null)
+            if (entities == null || entities.Count == 0)
             {
                 return null;
             }
@@ -148,8 +148,12 @@ namespace NLU.DevOps.Luis
                     .WithScore(score);
             }
 
-            var globalMetadata = entities["$instance"] as JToken;
-            if (globalMetadata == null)
+            var globalMetadata = default(JToken);
+            if (entities.TryGetValue("$instance", out var globalMetadataValue))
+            {
+                globalMetadata = globalMetadataValue as JToken;
+            }
+            else
             {
                 throw new InvalidOperationException("Expected top-level metadata for entities.");
             }

--- a/src/NLU.DevOps.LuisV3/LuisNLUTestClient.cs
+++ b/src/NLU.DevOps.LuisV3/LuisNLUTestClient.cs
@@ -149,9 +149,9 @@ namespace NLU.DevOps.Luis
             }
 
             var globalMetadata = default(JToken);
-            if (entities.TryGetValue("$instance", out var globalMetadataValue))
+            if (entities.TryGetValue("$instance", out var metadataValue) && metadataValue is JToken metadataJson)
             {
-                globalMetadata = globalMetadataValue as JToken;
+                globalMetadata = metadataJson;
             }
             else
             {

